plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/KGraphDiagramServer.xtend

```diff
@@ -642,6 +655,129 @@ class KGraphDiagramServer extends LanguageAwareDiagramServer {
         return
     }
     
+    protected def handle(RequestDiagramPieceAction action) {  
+        textsUpdated = false      
+        
+        synchronized (diagramState) {
+            if (diagramState.getDiagramPieceRequestManager(this.sourceUri) === null) {
+                // can't handle these requests if we are using recursive generation method
+                dispatch(new RejectAction())
+                return
+            }
+            val texts = diagramState.getTexts(this.sourceUri)
+            if (texts === null) {
+                throw new NullPointerException("The id of the SGraph was not found in the diagramState")
+            } else if (texts.empty) {
+                textsUpdated = true
+            } else {
+                val textDiagram = KGraphDiagramGenerator.generateTextDiagram(texts, this.sourceUri)
+                currentDiagramPieceRequest = action
+                dispatch(new IncrementalRequestTextBoundsAction(textDiagram))
+            }
+            // TODO: also handle images incrementally
+            if (textsUpdated) {
+                // go directly to next step
+                val diagramUpdater = diagramLanguageServer.diagramUpdater
+                if (diagramUpdater instanceof KGraphDiagramUpdater) {
+                    val piece = diagramUpdater.getNextDiagramPiece(this, action)
+                    if (piece !== null) {
+                        dispatch(new SetDiagramPieceAction(piece))
+                    } else {
+                        dispatch(new RejectAction())
+                    }
+                    
+                } else {
+                    throw new IllegalStateException("The diagramUpdater was not initialized correctly")
+                }
+            }            
+        }
+    }
+    
+    protected def handle(IncrementalComputedTextBoundsAction action) {
+                
+        synchronized (modelLock) {
+            /* 
+             * revision numbers lose their meaning a bit in this incremental concept so ignore this for now
```

Comment:
Well, the revision should be the same for all pieces, then the revision number still makes sense.